File: question/leetcode/leetcode0148.Sort-List/leetcode0148.go
package leetcode

import "clarencedu/algorithm-go/structure"

type ListNode = structure.ListNode

func sortList(head *ListNode) *ListNode {
	return mergeSort(head, nil)
}

// 归并
func mergeSort(head, tail *ListNode) *ListNode {
	if head == nil {
		return head
	}
	if head.Next == tail {
		head.Next = nil
		return head
	}
	slow, fast := head, head
	for fast != tail {
		slow = slow.Next
		fast = fast.Next
		if fast.Next != tail {
			fast = fast.Next
		}
	}
	mid := slow

	return mergeTwoList(mergeSort(head, mid), mergeSort(mid, tail))

}

func mergeTwoList(l1, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	current := dummy

	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			current.Next, l1 = l1, l1.Next
		} else {
			current.Next, l2 = l2, l2.Next
		}
		current = current.Next
	}

	if l1 != nil {
		current.Next = l1
	}
	if l2 != nil {
		current.Next = l2
	}

	return dummy.Next
}

func sortListII(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	mid := findMid(head)
	left := head
	right := mid.Next
	mid.Next = nil
	leftSort := sortList(left)
	rightSort := sortList(right)
	return mergeTwoListsII(leftSort, rightSort)
}

func findMid(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}
	if list2 == nil {
		return list1
	}
	if list1.Val < list2.Val {
		list1.Next = mergeTwoLists(list1.Next, list2)
		return list1
	} else {
		list2.Next = mergeTwoLists(list1, list2.Next)
	}
	return list2
}

func mergeTwoListsII(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	current := dummy
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			current.Next = list1
			list1 = list1.Next
		} else {
			current.Next = list2
			list2 = list2.Next
		}
		current = current.Next
	}

	if list1 != nil {
		current.Next = list1
	} else {
		current.Next = list2
	}

	return dummy.Next
}

File: question/leetcode/leetcode0148.Sort-List/leetcode0148.py
from typing import List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def mergeSort(head: ListNode, tail: ListNode) -> ListNode:
        if not head:
            return head
        if head.next == tail:
            head.next = None
            return head

        slow, fast = head, head
        while fast != tail:
            slow = slow.next
            fast = fast.next
            if fast and fast.next != tail:
                fast = fast.next

        mid = slow
        return mergeTwoLists(mergeSort(head, mid), mergeSort(mid, tail))

    def mergeTwoLists(l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode()
        current = dummy

        while l1 and l2:
            if l1.val < l2.val:
                current.next, l1 = l1, l1.next
            else:
                current.next, l2 = l2, l2.next
            current = current.next  # 别忘了移动 current 指针

        current.next = l1 if l1 else l2  # 直接连接剩余部分

        return dummy.next

    def sortList(head: ListNode) -> ListNode:
        return mergeSort(head, None)
